File: features/communication/chat_common/src/main/java/com/tokopedia/chat_common/util/ChatTimeConverter.java
package com.tokopedia.chat_common.util;

import android.content.Context;
import android.text.format.DateUtils;

import com.tokopedia.chat_common.R;

import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;

/**
 * Created by stevenfredian on 9/19/17.
 */

public class ChatTimeConverter {
    private static String TODAY = "Hari ini";
    private static String YESTERDAY = "Kemarin";

    public static Calendar unixToCalendar(long unixTime) {
        Calendar calendar = Calendar.getInstance();
        calendar.setTimeInMillis(unixTime);
        return calendar;
    }

    public static String formatdiff(long unixTime) {
        Locale localeID = new Locale("in", "ID");

        Date postTime = new Date(unixTime);

        Calendar calPostDate = Calendar.getInstance();
        calPostDate.setTime(postTime);

        Calendar calCurrentTime = Calendar.getInstance();
        calCurrentTime.setTime(new Date());

        if (DateUtils.isToday(unixTime)) {
            SimpleDateFormat sdfHour = new SimpleDateFormat("HH:mm", localeID);
            return sdfHour.format(postTime);
        } else if (calCurrentTime.get(Calendar.MONTH) == calPostDate.get(Calendar.MONTH)
                && calCurrentTime.get(Calendar.YEAR) == calPostDate.get(Calendar.YEAR)
                && isYesterday(calCurrentTime.get(Calendar.DAY_OF_MONTH), calPostDate.get(Calendar.DAY_OF_MONTH))) {
            return YESTERDAY;
        } else if (calCurrentTime.get(Calendar.YEAR) == calPostDate.get(Calendar.YEAR)) {
            SimpleDateFormat sdfDay = new SimpleDateFormat("dd MMM", localeID);
            return sdfDay.format(postTime);
        } else {
            SimpleDateFormat sdfYear = new SimpleDateFormat("dd MMM yyyy", localeID);
            return sdfYear.format(postTime);
        }
    }

    private static boolean isYesterday(int currentDay, int postDay) {
        return currentDay - postDay == 1;
    }

    public static String formatTime(long unixTime) {
        Locale localeID = new Locale("in", "ID");
        Date postTime = new Date(unixTime);
        SimpleDateFormat sdfHour = new SimpleDateFormat("HH:mm", localeID);
        return sdfHour.format(postTime);
    }

    public static String formatFullTime(long unixTime) {
        Locale localeID = new Locale("in", "ID");

        Date postTime = new Date(unixTime);

        Calendar calPostDate = Calendar.getInstance();
        calPostDate.setTime(postTime);

        Calendar calCurrentTime = Calendar.getInstance();
        calCurrentTime.setTime(new Date());

        if (DateUtils.isToday(unixTime)) {
            SimpleDateFormat sdfHour = new SimpleDateFormat("HH:mm", localeID);
            return (String.format("%s, %s", TODAY, sdfHour.format(postTime)));
        } else if (calCurrentTime.get(Calendar.MONTH) == calPostDate.get(Calendar.MONTH)
                && calCurrentTime.get(Calendar.YEAR) == calPostDate.get(Calendar.YEAR)
                && isYesterday(calCurrentTime.get(Calendar.DAY_OF_MONTH), calPostDate.get(Calendar.DAY_OF_MONTH))) {
            SimpleDateFormat sdfHour = new SimpleDateFormat("HH:mm", localeID);
            return (String.format("%s, %s", YESTERDAY, sdfHour.format(postTime)));
        } else if (calCurrentTime.get(Calendar.YEAR) == calPostDate.get(Calendar.YEAR)) {
            SimpleDateFormat sdfDay = new SimpleDateFormat("dd MMM, HH:mm", localeID);
            return sdfDay.format(postTime);
        } else {
            SimpleDateFormat sdfYear = new SimpleDateFormat("dd MMM yyyy, HH:mm", localeID);
            return sdfYear.format(postTime);
        }
    }

    public static String getRelativeDate(Context context, long unixTime) {
        long diff = (Calendar.getInstance().getTimeInMillis() / 1000) - unixTime;
        String status;
        long minuteDivider = 60;
        long hourDivider = minuteDivider * 60;
        long dayDivider = hourDivider * 24;
        long monthDivider = dayDivider * 30;
        if ((diff / monthDivider) > 0) {
            status = context.getString(R.string
                    .topchat_online_months_ago, diff / monthDivider);
        } else if ((diff / dayDivider) > 0) {
            status = context.getString(R.string
                    .topchat_online_days_ago, diff / dayDivider);
        } else if ((diff / hourDivider) > 0) {
            status = context.getString(R.string.topchat_online_hours_ago,
                    diff / hourDivider);
        } else {
            long minutes = diff / minuteDivider;
            if (minutes <= 0) minutes = 1;
            status = context.getString(R.string
                    .topchat_online_minutes_ago, minutes);
        }
        
        return status;
    }

}

File: features/communication/chat_common/src/main/java/com/tokopedia/chat_common/util/ChatTimeConverter.kt
package com.tokopedia.chat_common.util

import android.content.Context
import android.text.format.DateUtils
import com.tokopedia.chat_common.R
import java.text.SimpleDateFormat
import java.util.*

object ChatTimeConverter {

    private const val TODAY = "Hari ini"
    private const val YESTERDAY = "Kemarin"

    fun unixToCalendar(unixTime: Long): Calendar {
        val calendar = Calendar.getInstance()
        calendar.timeInMillis = unixTime
        return calendar
    }

    fun formatdiff(unixTime: Long): String {
        val localeID = Locale("in", "ID")
        val postTime = Date(unixTime)
        val calPostDate = Calendar.getInstance()
        calPostDate.time = postTime
        val calCurrentTime = Calendar.getInstance()
        calCurrentTime.time = Date()
        return if (DateUtils.isToday(unixTime)) {
            val sdfHour = SimpleDateFormat("HH:mm", localeID)
            sdfHour.format(postTime)
        } else if (calCurrentTime[Calendar.MONTH] == calPostDate[Calendar.MONTH] && calCurrentTime[Calendar.YEAR] == calPostDate[Calendar.YEAR] && isYesterday(
                calCurrentTime[Calendar.DAY_OF_MONTH],
                calPostDate[Calendar.DAY_OF_MONTH]
            )
        ) {
            YESTERDAY
        } else if (calCurrentTime[Calendar.YEAR] == calPostDate[Calendar.YEAR]) {
            val sdfDay = SimpleDateFormat("dd MMM", localeID)
            sdfDay.format(postTime)
        } else {
            val sdfYear = SimpleDateFormat("dd MMM yyyy", localeID)
            sdfYear.format(postTime)
        }
    }

    private fun isYesterday(currentDay: Int, postDay: Int): Boolean {
        return currentDay - postDay == 1
    }

    @JvmStatic
    fun formatTime(unixTime: Long): String {
        val localeID = Locale("in", "ID")
        val postTime = Date(unixTime)
        val sdfHour = SimpleDateFormat("HH:mm", localeID)
        return sdfHour.format(postTime)
    }

    fun formatFullTime(unixTime: Long): String {
        val localeID = Locale("in", "ID")
        val postTime = Date(unixTime)
        val calPostDate = Calendar.getInstance()
        calPostDate.time = postTime
        val calCurrentTime = Calendar.getInstance()
        calCurrentTime.time = Date()
        return if (DateUtils.isToday(unixTime)) {
            val sdfHour = SimpleDateFormat("HH:mm", localeID)
            String.format(
                "%s, %s",
                TODAY,
                sdfHour.format(postTime)
            )
        } else if (calCurrentTime[Calendar.MONTH] == calPostDate[Calendar.MONTH] && calCurrentTime[Calendar.YEAR] == calPostDate[Calendar.YEAR] && isYesterday(
                calCurrentTime[Calendar.DAY_OF_MONTH],
                calPostDate[Calendar.DAY_OF_MONTH]
            )
        ) {
            val sdfHour = SimpleDateFormat("HH:mm", localeID)
            String.format(
                "%s, %s",
                YESTERDAY,
                sdfHour.format(postTime)
            )
        } else if (calCurrentTime[Calendar.YEAR] == calPostDate[Calendar.YEAR]) {
            val sdfDay = SimpleDateFormat("dd MMM, HH:mm", localeID)
            sdfDay.format(postTime)
        } else {
            val sdfYear = SimpleDateFormat("dd MMM yyyy, HH:mm", localeID)
            sdfYear.format(postTime)
        }
    }

    fun getRelativeDate(context: Context, unixTime: Long): String {
        val diff = Calendar.getInstance().timeInMillis / 1000 - unixTime
        val status: String
        val minuteDivider: Long = 60
        val hourDivider = minuteDivider * 60
        val dayDivider = hourDivider * 24
        val monthDivider = dayDivider * 30
        when {
            diff / monthDivider > 0 -> {
                status = context.getString(R.string.topchat_online_months_ago, diff / monthDivider)
            }
            diff / dayDivider > 0 -> {
                status = context.getString(R.string.topchat_online_days_ago, diff / dayDivider)
            }
            diff / hourDivider > 0 -> {
                status = context.getString(
                    R.string.topchat_online_hours_ago,
                    diff / hourDivider
                )
            }
            else -> {
                var minutes = diff / minuteDivider
                if (minutes <= 0) minutes = 1
                status = context.getString(R.string.topchat_online_minutes_ago, minutes)
            }
        }
        return status
    }
}